lib/FrontendTool/FrontendTool.cpp

```diff
@@ -685,6 +685,22 @@ static SourceFile *getPrimaryOrMainSourceFile(CompilerInvocation &Invocation,
   return SF;
 }
 
+/// Dumps the AST of all available primary source files. If corresponding output
+/// files were specified, use them; otherwise, dump the AST to stderr.
+static void dumpAST(CompilerInstance &Instance) {
+  for(SourceFile *sourceFile: Instance.getPrimarySourceFiles()) {
+    const StringRef OutputFilename =
+      Instance.getPrimarySpecificPathsForSourceFile(*sourceFile).OutputFilename;
+    // PSP.OutputFilename is "-" whenever the output should be written to the console.
```

Comment:
I think `getFileOutputStream` handles this, actually, by using `llvm::raw_fd_ostream`'s constructor. So you can use that.

Reply:
It seems that `llvm::raw_fd_ostream`'s constructor defaults to `stdout` when the filename is `"-"`, whereas the AST dump defaults to `stderr`.
I personally wouldn't mind changing the behavior to use `stdout` here, as it would slightly simplify both the code and my research... but maybe it's standard practice to use `stderr` for these things? I don't know.

Reply:
It's my understanding that we actually can't dump the ast at the same time as another output mode (which is why I have to remove the `-emit-...` options from the Xcode command), so yeah, I think it should be fine.

Reply:
It seems that dropping this makes the default constructor return a reference to `stdout` that isn't `llvm::outs()`. This causes the AST dumping code to dump everything without colors, since it checks for `&os == &llvm::outs()`, which fails. Do you know another way to check it the OS is `stdout`? Otherwise, I think I'm just gonna leave the manual check and default to `llvm::errs()` to preserve the colors.

Reply:
Arguably that `has_colors` check is wrong. I don't see why we're not just asking the stream itself.

Reply:
Did you want to do something about this for this PR, or follow up after?

Reply:
I fixed it in my last commit! I thought it’d show up here, maybe I’m doing git wrong heh sorry

Reply:
Ah, but that check is copy/pasted all over ASTDumper…

Reply:
Oh, I hadn’t noticed... but yeah, I think I’d better do it in this PR. I’ll send an update tomorrow then.

Reply:
The other checks were setting `ShowColors` attributes in some (but not all) ast dump classes, but these attributes were never used again.
I removed the attributes (and the checks), so now we only do the check in one place, and it's done "correctly", and the classes are more consistent.

Reply:
Yeah, for the record I'm having trouble building master too, so I think this is OK. Let's see what the CI says.